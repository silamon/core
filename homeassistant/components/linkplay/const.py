"""LinkPlay constants."""

from homeassistant.const import Platform

DOMAIN = "linkplay"
CONTROLLER = "controller"
PLATFORMS = [Platform.MEDIA_PLAYER]
DATA_SESSION = "session"
