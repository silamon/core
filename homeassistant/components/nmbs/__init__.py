"""The NMBS component."""

import logging

from homeassistant.config_entries import SOURCE_IMPORT, ConfigEntry
from homeassistant.const import CONF_PLATFORM, CONF_SHOW_ON_MAP, CONF_TYPE, Platform
from homeassistant.core import DOMAIN as HOMEASSISTANT_DOMAIN, HomeAssistant
from homeassistant.helpers.issue_registry import IssueSeverity, async_create_issue
from homeassistant.helpers.typing import ConfigType

from .const import CONF_EXCLUDE_VIAS, CONF_STATION_LIVE, DOMAIN  # noqa: F401

_LOGGER = logging.getLogger(__name__)
PLATFORMS = [Platform.SENSOR]


async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
    """Set up the NMBS component through yaml (deprecated)."""

    platforms_config: dict[Platform, list[ConfigType]] = {
        domain: config[domain] for domain in PLATFORMS if domain in config
    }
    for items in platforms_config.values():
        for item in items:
            if item[CONF_PLATFORM] == DOMAIN:
                if CONF_SHOW_ON_MAP not in item:
                    item[CONF_SHOW_ON_MAP] = False
                if CONF_EXCLUDE_VIAS not in item:
                    item[CONF_EXCLUDE_VIAS] = False

                item[CONF_TYPE] = "connection"
                hass.async_create_task(
                    hass.config_entries.flow.async_init(
                        DOMAIN,
                        context={"source": SOURCE_IMPORT},
                        data=item,
                    )
                )
                if CONF_STATION_LIVE in item:
                    liveboard = item.copy()
                    liveboard[CONF_TYPE] = "liveboard"
                    hass.async_create_task(
                        hass.config_entries.flow.async_init(
                            DOMAIN,
                            context={"source": SOURCE_IMPORT},
                            data=liveboard,
                        )
                    )

    async_create_issue(
        hass,
        HOMEASSISTANT_DOMAIN,
        f"deprecated_yaml_{DOMAIN}",
        breaks_in_ha_version="2025.3.0",
        is_fixable=False,
        issue_domain=DOMAIN,
        severity=IssueSeverity.WARNING,
        translation_key="deprecated_yaml",
        translation_placeholders={
            "domain": DOMAIN,
            "integration_title": "NMBS",
        },
    )

    return True


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up NMBS from a config entry."""

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
    return True


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
