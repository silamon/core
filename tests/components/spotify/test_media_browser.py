"""Test the media browser interface."""

from unittest.mock import MagicMock

import pytest
from syrupy import SnapshotAssertion

from homeassistant.components.spotify import DOMAIN
from homeassistant.components.spotify.browse_media import async_browse_media
from homeassistant.const import CONF_ID
from homeassistant.core import HomeAssistant

from . import setup_integration
from .conftest import SCOPES

from tests.common import MockConfigEntry


@pytest.mark.usefixtures("setup_credentials")
async def test_browse_media_root(
    hass: HomeAssistant,
    mock_spotify: MagicMock,
    mock_config_entry: MockConfigEntry,
    snapshot: SnapshotAssertion,
    expires_at: int,
) -> None:
    """Test browsing the root."""
    await setup_integration(hass, mock_config_entry)
    # We add a second config entry to test that lowercase entry_ids also work
    config_entry = MockConfigEntry(
        domain=DOMAIN,
        title="spotify_2",
        unique_id="second_fake_id",
        data={
            CONF_ID: "second_fake_id",
            "name": "spotify_account_2",
            "auth_implementation": DOMAIN,
            "token": {
                "access_token": "mock-access-token",
                "refresh_token": "mock-refresh-token",
                "expires_at": expires_at,
                "scope": SCOPES,
            },
        },
        entry_id="32oesphrnacjcf7vw5bf6odx3",
    )
    await setup_integration(hass, config_entry)
    response = await async_browse_media(hass, None, None)
    assert response.as_dict() == snapshot


@pytest.mark.usefixtures("setup_credentials")
async def test_browse_media_categories(
    hass: HomeAssistant,
    mock_spotify: MagicMock,
    mock_config_entry: MockConfigEntry,
    snapshot: SnapshotAssertion,
) -> None:
    """Test browsing categories."""
    await setup_integration(hass, mock_config_entry)
    response = await async_browse_media(
        hass, "spotify://library", f"spotify://{mock_config_entry.entry_id}"
    )
    assert response.as_dict() == snapshot


@pytest.mark.parametrize(
    ("config_entry_id"), [("01J5TX5A0FF6G5V0QJX6HBC94T"), ("32oesphrnacjcf7vw5bf6odx3")]
)
@pytest.mark.usefixtures("setup_credentials")
async def test_browse_media_playlists(
    hass: HomeAssistant,
    config_entry_id: str,
    mock_spotify: MagicMock,
    snapshot: SnapshotAssertion,
    expires_at: int,
) -> None:
    """Test browsing playlists for the two config entries."""
    mock_config_entry = MockConfigEntry(
        domain=DOMAIN,
        title="Spotify",
        unique_id="1112264649",
        data={
            "auth_implementation": DOMAIN,
            "token": {
                "access_token": "mock-access-token",
                "refresh_token": "mock-refresh-token",
                "expires_at": expires_at,
                "scope": SCOPES,
            },
        },
        entry_id=config_entry_id,
    )
    await setup_integration(hass, mock_config_entry)
    response = await async_browse_media(
        hass,
        "spotify://current_user_playlists",
        f"spotify://{config_entry_id}/current_user_playlists",
    )
    assert response.as_dict() == snapshot
