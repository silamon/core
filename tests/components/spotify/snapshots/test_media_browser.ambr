# serializer version: 1
# name: test_browse_media_categories
  dict({
    'can_expand': True,
    'can_play': False,
    'children': list([
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': <MediaClass.PLAYLIST: 'playlist'>,
        'media_class': <MediaClass.DIRECTORY: 'directory'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/current_user_playlists',
        'media_content_type': 'spotify://current_user_playlists',
        'thumbnail': None,
        'title': 'Playlists',
      }),
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': <MediaClass.ARTIST: 'artist'>,
        'media_class': <MediaClass.DIRECTORY: 'directory'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/current_user_followed_artists',
        'media_content_type': 'spotify://current_user_followed_artists',
        'thumbnail': None,
        'title': 'Artists',
      }),
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': <MediaClass.ALBUM: 'album'>,
        'media_class': <MediaClass.DIRECTORY: 'directory'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/current_user_saved_albums',
        'media_content_type': 'spotify://current_user_saved_albums',
        'thumbnail': None,
        'title': 'Albums',
      }),
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': <MediaClass.TRACK: 'track'>,
        'media_class': <MediaClass.DIRECTORY: 'directory'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/current_user_saved_tracks',
        'media_content_type': 'spotify://current_user_saved_tracks',
        'thumbnail': None,
        'title': 'Tracks',
      }),
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': <MediaClass.PODCAST: 'podcast'>,
        'media_class': <MediaClass.DIRECTORY: 'directory'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/current_user_saved_shows',
        'media_content_type': 'spotify://current_user_saved_shows',
        'thumbnail': None,
        'title': 'Podcasts',
      }),
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': <MediaClass.TRACK: 'track'>,
        'media_class': <MediaClass.DIRECTORY: 'directory'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/current_user_recently_played',
        'media_content_type': 'spotify://current_user_recently_played',
        'thumbnail': None,
        'title': 'Recently played',
      }),
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': <MediaClass.ARTIST: 'artist'>,
        'media_class': <MediaClass.DIRECTORY: 'directory'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/current_user_top_artists',
        'media_content_type': 'spotify://current_user_top_artists',
        'thumbnail': None,
        'title': 'Top Artists',
      }),
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': <MediaClass.TRACK: 'track'>,
        'media_class': <MediaClass.DIRECTORY: 'directory'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/current_user_top_tracks',
        'media_content_type': 'spotify://current_user_top_tracks',
        'thumbnail': None,
        'title': 'Top Tracks',
      }),
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': <MediaClass.GENRE: 'genre'>,
        'media_class': <MediaClass.DIRECTORY: 'directory'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/categories',
        'media_content_type': 'spotify://categories',
        'thumbnail': None,
        'title': 'Categories',
      }),
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': <MediaClass.PLAYLIST: 'playlist'>,
        'media_class': <MediaClass.DIRECTORY: 'directory'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/featured_playlists',
        'media_content_type': 'spotify://featured_playlists',
        'thumbnail': None,
        'title': 'Featured Playlists',
      }),
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': <MediaClass.ALBUM: 'album'>,
        'media_class': <MediaClass.DIRECTORY: 'directory'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/new_releases',
        'media_content_type': 'spotify://new_releases',
        'thumbnail': None,
        'title': 'New Releases',
      }),
    ]),
    'children_media_class': <MediaClass.DIRECTORY: 'directory'>,
    'media_class': <MediaClass.DIRECTORY: 'directory'>,
    'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/library',
    'media_content_type': 'spotify://library',
    'not_shown': 0,
    'thumbnail': None,
    'title': 'Media Library',
  })
# ---
# name: test_browse_media_playlists[01J5TX5A0FF6G5V0QJX6HBC94T]
  dict({
    'can_expand': True,
    'can_play': False,
    'children': list([
      dict({
        'can_expand': True,
        'can_play': True,
        'children_media_class': <MediaClass.TRACK: 'track'>,
        'media_class': <MediaClass.PLAYLIST: 'playlist'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/spotify:playlist:4WkWJ0EjHEFASDevhM8oPw',
        'media_content_type': 'spotify://playlist',
        'thumbnail': 'https://i.scdn.co/image/ab67616d0000b273d061f5bfae8d38558f3698c1',
        'title': 'Hyper',
      }),
      dict({
        'can_expand': True,
        'can_play': True,
        'children_media_class': <MediaClass.TRACK: 'track'>,
        'media_class': <MediaClass.PLAYLIST: 'playlist'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/spotify:playlist:1RHirWgH1weMsBLi4KOK9d',
        'media_content_type': 'spotify://playlist',
        'thumbnail': 'https://mosaic.scdn.co/640/ab67616d0000b2732f3e58dd611d177973cb3a8cab67616d0000b27345cab965cb4639a4e669564aab67616d0000b2739e83c93811be6abfad8649d6ab67616d0000b273e4c03429788f0aff263a5fc6',
        'title': 'Ain’t got shit on me',
      }),
    ]),
    'children_media_class': <MediaClass.PLAYLIST: 'playlist'>,
    'media_class': <MediaClass.DIRECTORY: 'directory'>,
    'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/current_user_playlists',
    'media_content_type': 'spotify://current_user_playlists',
    'not_shown': 0,
    'thumbnail': None,
    'title': 'Playlists',
  })
# ---
# name: test_browse_media_playlists[32oesphrnacjcf7vw5bf6odx3]
  dict({
    'can_expand': True,
    'can_play': False,
    'children': list([
      dict({
        'can_expand': True,
        'can_play': True,
        'children_media_class': <MediaClass.TRACK: 'track'>,
        'media_class': <MediaClass.PLAYLIST: 'playlist'>,
        'media_content_id': 'spotify://32oesphrnacjcf7vw5bf6odx3/spotify:playlist:4WkWJ0EjHEFASDevhM8oPw',
        'media_content_type': 'spotify://playlist',
        'thumbnail': 'https://i.scdn.co/image/ab67616d0000b273d061f5bfae8d38558f3698c1',
        'title': 'Hyper',
      }),
      dict({
        'can_expand': True,
        'can_play': True,
        'children_media_class': <MediaClass.TRACK: 'track'>,
        'media_class': <MediaClass.PLAYLIST: 'playlist'>,
        'media_content_id': 'spotify://32oesphrnacjcf7vw5bf6odx3/spotify:playlist:1RHirWgH1weMsBLi4KOK9d',
        'media_content_type': 'spotify://playlist',
        'thumbnail': 'https://mosaic.scdn.co/640/ab67616d0000b2732f3e58dd611d177973cb3a8cab67616d0000b27345cab965cb4639a4e669564aab67616d0000b2739e83c93811be6abfad8649d6ab67616d0000b273e4c03429788f0aff263a5fc6',
        'title': 'Ain’t got shit on me',
      }),
    ]),
    'children_media_class': <MediaClass.PLAYLIST: 'playlist'>,
    'media_class': <MediaClass.DIRECTORY: 'directory'>,
    'media_content_id': 'spotify://32oesphrnacjcf7vw5bf6odx3/current_user_playlists',
    'media_content_type': 'spotify://current_user_playlists',
    'not_shown': 0,
    'thumbnail': None,
    'title': 'Playlists',
  })
# ---
# name: test_browse_media_root
  dict({
    'can_expand': True,
    'can_play': False,
    'children': list([
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': None,
        'media_class': <MediaClass.APP: 'app'>,
        'media_content_id': 'spotify://01J5TX5A0FF6G5V0QJX6HBC94T',
        'media_content_type': 'spotify://library',
        'thumbnail': 'https://brands.home-assistant.io/_/spotify/logo.png',
        'title': 'spotify_1',
      }),
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': None,
        'media_class': <MediaClass.APP: 'app'>,
        'media_content_id': 'spotify://32oesphrnacjcf7vw5bf6odx3',
        'media_content_type': 'spotify://library',
        'thumbnail': 'https://brands.home-assistant.io/_/spotify/logo.png',
        'title': 'spotify_2',
      }),
    ]),
    'children_media_class': <MediaClass.APP: 'app'>,
    'media_class': <MediaClass.APP: 'app'>,
    'media_content_id': 'spotify://',
    'media_content_type': 'spotify',
    'not_shown': 0,
    'thumbnail': 'https://brands.home-assistant.io/_/spotify/logo.png',
    'title': 'Spotify',
  })
# ---
